include/h2o/httpclient.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2017 Ichito Nagata, Fastly, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#ifndef h2o__httpclient_h
+#define h2o__httpclient_h
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "h2o/memory.h"
+#include "h2o/socket.h"
+#include "h2o/socketpool.h"
+#include "h2o/timeout.h"
+#include "h2o/header.h"
+
+typedef struct st_h2o_httpclient_t h2o_httpclient_t;
+
+typedef struct st_h2o_httpclient_features_t {
+    h2o_iovec_t *proxy_protocol;
+    int *chunked;
+    int connection_header;
+} h2o_httpclient_features_t;
+
+struct st_h2o_header_t;
+
+typedef void (*h2o_httpclient_proceed_req_cb)(h2o_httpclient_t *client, size_t written, int is_end_stream);
+typedef int (*h2o_httpclient_body_cb)(h2o_httpclient_t *client, const char *errstr);
+typedef h2o_httpclient_body_cb (*h2o_httpclient_head_cb)(h2o_httpclient_t *client, const char *errstr, int minor_version,
+                                                           int status, h2o_iovec_t msg, struct st_h2o_header_t *headers,
+                                                           size_t num_headers, int rlen);
+typedef h2o_httpclient_head_cb (*h2o_httpclient_connect_cb)(h2o_httpclient_t *client, const char *errstr, h2o_iovec_t *method, h2o_url_t *url,
+                                                              h2o_headers_t *headers, h2o_iovec_t *body,
```

Comment:
Considering the fact that we will be passing back a _constant_ list of headers, the argument should be `const h2o_header_t **headers, size_t *num_headers` instead of `h2o_headers_t *headers`.